pages/blog/subscriptions-and-live-queries-real-time-with-graphql.mdx

```diff
@@ -0,0 +1,1150 @@
+
+When using services like PostGraphile and Hasura that is the case. However, for any additional resolvers that are added on top of the service schema, we cannot implement an invalidation mechanism.
+In user-land we will have to come up with an implementation of resource tracking by ourselves.
+
+2. Efficiently updates clients. **No.** ❌
+
+The whole execution result is sent over to the client for every single time a live query is invalidated.
+
+3. Flexibility. **Low.** ❌
+
+Both Hasura and PostGraphile are tightly coupled to a SQL database. For any custom resolvers, we have to come up with the mechanism for resource tracking and invalidation ourselves.
+
+4. Adoptability. **Low.** ❌
+
+Switching to a server powered by PostGraphile or Hasura with an already existing GraphQL schema is no easy task.
+
+## GraphQL Live Queries over Subscriptions with JSON patch
+
+Ideally, we only want to send a patch to the client that provides instructions on how to get from the previous execution result to the next. The lack of these instructions has been a big flaw in the previous two implementations.
+
+The RFCs and implementations for the `@defer` and `@stream` introduced ways of sending (delayed) partial results to clients. However, those "patch" operations are currently highly limited to a "replace at path" and "append to list" operation.
+
+A format such as [JSON Patch](https://tools.ietf.org/html/rfc6902) might be a better alternative for live queries.
+
+[`graphql-live-subscriptions`](https://github.com/D1plo1d/graphql-live-subscriptions) tries to solve that with `Subscription.live` field that exposes both a `Query` and a JSON patch field.
+
+**Schema Types for `graphql-live-subscriptions`**
+
+```graphql
+scalar JSON
+
+type RFC6902Operation {
+    op: String!
+    path: String!
+    from: String
+    value: JSON
+  }
+
+type Subscription {
+  live: {
+    patch: [RFC6902Operation]
+    query: Query
+  }
+}
+```
+
+A live query operation can be declared similar to our `PostSubscription` document above.
+
+```ts
+const PostSubscription = graphql`
+  subscription PostSubscription($postId: ID!, $userId: ID!) {
+    live {
+      query {
+        post(id: $postId) {
+          id
+          title
+          totalLikeCount
+        }
+        user(id: $userId) {
+          id
+          name
+        }
+      }
+      patch {
+        op
+        path
+        from
+        value
+      }
+    }
+  }
+`;
+```
+
+The difference is that the type returned by the `live` field has two fields instead of a single one.
+The `query` field, which selects the selection set from `Query` tyoe and a `patch` field which is a JSON Patch operation.
+When executing the given operation against the server the initial result will have the data selected by the `query` field selection set included.
+All following values, will have no `query` value (`null`), but instead a array of patch operations that describe the changes for updating the last result to the next result.
```

Comment:
Just a small grammatical correction: It should be `an array`

Reply:
```suggestion
All following values will have no `query` value (`null`), but instead an array of patch operations that describe the changes for updating the last result to the next result.
```